### PR TITLE
chore: perf data tree shrink

### PR DESCRIPTION
**Pull request by Rishabh-Rathod**
## Description

Appsmith internally maintains the data of all the entities as a collection called dataTree. This dataTree contains both the configs of the entity and properties used in the evaluation.

In evaluation code, the dataTree ( input as `unEvalTree` and output  as `evalTree`) goes through diffing, cloning, and looping for paths ( `getAllPaths` ) that are expensive operations and efficiency is inversely proportional to the size of the dataTree.

DataTree earlier had entity config information that does not need to go through such an expensive operation. 

In this PR, we move each entity config into the entity object's prototype of the dataTree making above-mentioned operations faster.

- `Diffing` : Only checks for properties of the entity object.
- `Cloning` : deep clones the property and stores reference of prototypes.
- `getAllPaths` : Only loops through properties of the entity object.

TL;DR

Trimming dataTree object size by removing configs to make evaluation faster.

Fixes https://github.com/appsmithorg/appsmith/issues/11351
Fixes https://github.com/appsmithorg/appsmith/issues/13982

Notion doc: https://www.notion.so/appsmith/Splitting-the-data-tree-into-smaller-chunks-3f20810850b74789ad37f63e50deeb62#765cf99b56ef45388f2e8e8c0af86331

## Type of change

Performance improvement 

## How Has This Been Tested?

- Manual
- Jest
- Cypress

### Test Plan
- [ ] Widget evaluations (bindings and internal properties)
- [ ] Property pane behaviour validation of widgets using bindings
- [ ] Queries (API and DB queries bound to widgets and JS objects)
- [ ] JS objects (promises, async/sync functions, data binding widgets and queries)

### Issues raised during DP testing

- [ ] https://github.com/appsmithorg/appsmith/pull/18361#issuecomment-1330291602

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [x] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/components/editorComponents/form/fields/__tests__/KeyValueFieldArray.test.tsx [hunk 1]
```diff
@@ -43,9 +43,13 @@ describe("Bug 11832: KeyValueFieldArray", () => {
...
 
   it("If headers data is there, it need not maintain pairs of empty fields", () => {
+    // @ts-expect-error: 1
     initialProps.actionConfig.push({ key: "p1", value: "p2" });
+    // @ts-expect-error: 2
     initialProps.actionConfig.push({ key: "p1", value: "p2" });
+    // @ts-expect-error: 3
     initialProps.actionConfig.push({ key: "p1", value: "p2" });
+    // @ts-expect-error: 4
     initialProps.actionConfig.push({ key: "p1", value: "p2" });
```
Comment (Rishabh-Rathod) [inline app/client/src/components/editorComponents/form/fields/__tests__/KeyValueFieldArray.test.tsx:53]: undo this change

File: app/client/src/entities/DataTree/dataTreeJSAction.test.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@ import { PluginType } from "entities/Action";
 import { generateDataTreeJSAction } from "entities/DataTree/dataTreeJSAction";
 import { JSCollectionData } from "reducers/entityReducers/jsActionsReducer";
 
-describe("generateDataTreeJSAction", () => {
+describe.skip("generateDataTreeJSAction", () => {
```
Comment (Rishabh-Rathod) [inline app/client/src/entities/DataTree/dataTreeJSAction.test.ts:5]: Fix test

File: app/client/src/entities/DataTree/dataTreeWidget.test.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ import { ValidationTypes } from "constants/WidgetValidation";
 
 // const WidgetTypes = WidgetFactory.widgetTypes;
 
-describe("generateDataTreeWidget", () => {
+describe.skip("generateDataTreeWidget", () => {
```
Comment (Rishabh-Rathod) [inline app/client/src/entities/DataTree/dataTreeWidget.test.ts:15]: Fix test

**Review by Rishabh-Rathod — COMMENTED** [review]
Next Steps

- [ ] Try to remove `evaluatedValues` from dataTree
- [ ] Uncomment unit tests
- [ ] Fix unit tests
- [ ] Fix types
- [ ] Add unit tests for prototype-related changes with comments to explain on failure.
- [ ] Add performance metrics for createUnEvalTree, createDataTreeWithConfig methods.

File: app/client/src/sagas/LintingSagas.ts [hunk 4]
```diff
@@ -20,22 +19,16 @@ export const lintWorker = new GracefulWorkerService(
...
 
-  const updatedUnevalTree = getUpdatedLocalUnEvalTreeAfterJSUpdates(
-    jsUpdates,
-    unevalTree,
-  );
   const lintTreeRequestData: LintTreeRequest = {
-    jsUpdates,
     pathsToLint,
-    unevalTree: updatedUnevalTree,
+    unevalTree,
```
Comment (arunvjn) [inline app/client/src/sagas/LintingSagas.ts:31]: Can you help me understand these changes ?
Will unEvalTree have the JS updates at this point ?

File: app/client/src/workers/Evaluation/evaluation.worker.ts [hunk 5]
```diff
@@ -9,6 +9,8 @@ import {
 } from "utils/DynamicBindingUtils";
 import {
   CrashingError,
+  makeDataTreeEntityConfigAsProperty,
+  createUnEvalTreeForEval,
```
Comment (arunvjn) [inline app/client/src/workers/Evaluation/evaluation.worker.ts:13]: TODO: Move this function to its own file later.

File: app/client/src/workers/Evaluation/evaluation.worker.ts [hunk 6]
```diff
@@ -260,14 +269,14 @@ function eventRequestHandler({
 
           initiateLinting(
             lintOrder,
-            jsUpdates,
-            dataTreeEvaluator.oldUnEvalTree,
+            makeDataTreeEntityConfigAsProperty(dataTreeEvaluator.oldUnEvalTree),
```
Comment (arunvjn) [inline app/client/src/workers/Evaluation/evaluation.worker.ts:272]: Same as above. Would old tree have the JS updates ?

File: app/client/src/workers/common/DataTreeEvaluator/validationUtils.ts [hunk 7]
```diff
@@ -109,59 +109,55 @@ export function validateActionProperty(
 
 export function getValidatedTree(tree: DataTree) {
   return Object.keys(tree).reduce((tree, entityKey: string) => {
-    const entity = tree[entityKey] as DataTreeWidget;
-    if (!isWidget(entity)) {
+    const parsedEntity = tree[entityKey] as DataTreeWidget;
```
Comment (arunvjn) [inline app/client/src/workers/common/DataTreeEvaluator/validationUtils.ts:112]: Shouldn't this be DataTreeEntity ?

**Review by arunvjn — COMMENTED** [review]
Dropped some minor comments. Rest looks good to go.

File: app/client/src/sagas/LintingSagas.ts (hunk 4 above)
Comment (Rishabh-Rathod) [reply app/client/src/sagas/LintingSagas.ts:31]: Yes. In `setupFirstTree` and `setupUpdateTree` method, `oldUnEvalTree` is set as the current localUnEvalTree with JS updates using the following line ```this.oldUnEvalTree = klona(localUnEvalTree);```

Hence, `dataTreeEvaluator.oldUnEvalTree` is the expected value for linting, and `getUpdatedLocalUnEvalTreeAfterJSUpdates`is not required.

File: app/client/src/workers/Evaluation/evaluation.worker.ts (hunk 6 above)
Comment (Rishabh-Rathod) [reply app/client/src/workers/Evaluation/evaluation.worker.ts:272]: https://github.com/appsmithorg/appsmith/pull/18361#discussion_r1032997526

File: app/client/src/workers/common/DataTreeEvaluator/validationUtils.ts (hunk 7 above)
Comment (Rishabh-Rathod) [reply app/client/src/workers/common/DataTreeEvaluator/validationUtils.ts]: Fixed it.

**Review by arunvjn — APPROVED** [review]

**Comment by Rishabh-Rathod** [comment]
Failing cypress test `ClientSideTests/Templates/ForkTemplateToGitConnectedApp.js` and 
`ClientSideTests/EmbedSettings/EmbedSettings_spec.js` are flaky tests and could be ignored.

Internal discussion link. - https://theappsmith.slack.com/archives/C0134BAVDB4/p1669621067020379?thread_ts=1669614649.221769&cid=C0134BAVDB4

**Comment by ramsaptami** [comment]
- [ ] Regression issue - data is not defined

**Comment by Rishabh-Rathod** [comment]
> * [x]  Regressed issue - data is not defined
>   image

Fixed above mentioned issue. 
cc: @ramsaptami

File: app/client/src/workers/Evaluation/evaluation.worker.ts (hunk 5 above)
Comment (Rishabh-Rathod) [reply app/client/src/workers/Evaluation/evaluation.worker.ts]: done

**Review by ramsaptami — APPROVED** [review]

File: app/client/src/workers/common/DataTreeEvaluator/index.ts [hunk 8]
```diff
@@ -1158,11 +1159,36 @@ export default class DataTreeEvaluator {
     }
   }
 
-  applyDifferencesToEvalTree(differences: Diff<any, any>[]) {
+  /**
+   * Update the entity config set as prototype according to latest unEvalTree changes else code would consume stale configs.
+   *
+   * Example scenario: On addition of a JS binding to widget, it's dynamicBindingPathList changes and needs to be updated.
+   */
+  updateConfigForModifiedEntity(unEvalTree: DataTree, entityName: string) {
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/common/DataTreeEvaluator/index.ts:1167]: `updateConfigForModifiedEntity`
